lib/core/src/Cardano/Wallet/Primitive/Fee.hs

```diff
@@ -213,16 +213,33 @@ coverRemainingFee (Fee fee) = go [] where
 -- change outputs if any are left, or the remaining fee otherwise
 --
 -- We divvy up the fee over all change outputs proportionally, to try and keep
--- any output:change ratio as unchanged as possible
-rebalanceChangeOutputs
+-- any output:change ratio as unchanged as possible.
+--
+-- This function either consumes an existing reserve on a selection, or turn it
+-- into a change output. Therefore, the resulting coin selection _will_ not have
+-- any reserve. Note that the reserve will be either 'Nothing', to indicate that
+-- there was no reserve at all, or 'Just 0' to indicate that there was a
+-- reserve, but it has been consumed entirely.
+rebalanceSelection
     :: FeeOptions
     -> CoinSelection
     -> (CoinSelection, Fee)
-rebalanceChangeOutputs opts s
+rebalanceSelection opts s
     -- selection is now balanced, nothing to do.
     | φ_original == δ_original =
         (s, Fee 0)
 
+    -- some fee left to pay, and the reserve is non-empty, use it first.
+    | φ_original > δ_original && reserve s > Just (Coin 0) =
+        let
+            -- Safe because of the above guard.
+            Just (Coin r) = reserve s
+            r' = if r > φ_original
+                then Coin (r - φ_original)
```

Comment:
it was supposed to turn all reserve (if there is left) into change. So `rebalanceSelection` will be called in with decreased reserve and some branch will decrease it to zero and after that it will return in `senderPaysFee` with reserve (Just (Coin 0)). Is that right?

Reply:
Correct.

Reply:
Hence the added check in the corresponding property checking that the reserve is always 0 at the end of the balancing.